src/node/routes/login.ts

```diff
@@ -87,13 +83,13 @@ router.post("/", async (req, res) => {
     if (isPasswordValid) {
       // The hash does not add any actual security but we do it for
       // obfuscation purposes (and as a side effect it handles escaping).
-      res.cookie(Cookie.Key, hashedPassword, {
+      res.cookie(Cookie.SessionKey, hashedPassword, {
         domain: getCookieDomain(req.headers.host || "", req.args["proxy-domain"]),
         // Browsers do not appear to allow cookies to be set relatively so we
         // need to get the root path from the browser since the proxy rewrites
         // it out of the path.  Otherwise code-server instances hosted on
         // separate sub-paths will clobber each other.
-        path: req.body.base ? path.posix.join(req.body.base, "..") : "/",
+        path: req.body.base ? path.posix.join(req.body.base, "..", "/") : "/",
```

Comment:
Previously this would assign the cookie to a resource path instead of a route path (e.g. `/code` instead of `/code/`). From what I’ve researched, there’s a [number of opinions](https://stackoverflow.com/questions/5948659/when-should-i-use-a-trailing-slash-in-my-url) in [both camps](https://cdivilly.wordpress.com/2014/03/11/why-trailing-slashes-on-uris-are-important/). I believe that since we’re now leaning so heavily on relative routes in VS Code, there’s little we can do to avoid this aside from convenient redirects and providing guidance for users.

- code-server may be accessed either at the root or at /vscode with no slash
- If you’re using a proxy that includes a path, it must end in a slash
- If you’d like to omit the slash with your proxy, this must be configured on your end

Reply:
> If you’d like to omit the slash with your proxy, this must be configured on your end

I feel like this is a fair tradeoff but I didn't use proxies with code-server before so can't fully speak to that UX. @code-asher what are your thoughts?

Reply:
The issue with omitting the trailing slash from the cookie is that logging out from any arbitrary path requires that we send that very same path in the logout request. Trailing slashes denote the depth and access of a cookie, and while the spec is pretty forgiving on cookies to omit it, we can’t identify what part of the URL is where authenticated “starts” without knowing if the user first authenticated from `/ ` or `/code`. Both share the same path base `/` while adding a trailing slash defines cookies from `/code/`.

Reply:
Ah so a cookie set to the path `/code` actually scopes the cookie to `/`?  That does seem like a good reason to add the trailing slash.

Reply:
Hmm the docs here suggest that is not the case: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#define_where_cookies_are_sent

To quote the relevant portion:

For example, if you set Path=/docs, these request paths match:

    /docs
    /docs/
    /docs/Web/
    /docs/Web/HTTP

But these request paths don't:

    /
    /docsets
    /fr/docs

Reply:
That’s correct, the spec lets the trailing slash be omitted however we still can’t know which depth they’re attempting to login or out of without a trailing slash since the URL spec is more strict.

Reply:
We need the depth either way right?

For example assuming the base path is `/apps/coder/code-server/` when we log out from `/apps/coder/code-server/vscode/` we now get redirected to `/apps/coder/code-server/logout?base=/apps/coder/code-server/vscode/` and from there we need to set the cookie path using `path.posix.join(base, "..")` or `path.posix.join(base, "..", "/")` (only difference is the trailing slash).

If we log out from `/apps/coder/code-server/` then we need to do the same thing in both cases except without a `..`.  If there was some future route like `/apps/coder/code-server/my/new/route/` we would need `../../..` in both cases.

Maybe there is some way we can get rid of needing to know what the current depth is?  Hmmmm... :thinking: 

In previous versions of code-server the way we made this work was to pass the relative path to the root to the frontend (so in the three examples above it would be `..`, `.`, and `../../..` respectively).  The client then resolved this against its current URL (in all three cases it would resolve to `/apps/coder/code-server`) and then we would redirect to `logout?base=/apps/coder/code-server` and we could set the path directly to that query variable since it was already resolved to the root.

Reply:
Maybe where we pass things like `logoutEndpoint` we can also include the current route so we can then derive the correct base.

Reply:
To summarize, @code-asher you're saying this change would break current behavior?

I think you and I will need to test and verify this ourselves. If it does lead to a breaking change, we should see what our options are like you said.

Reply:
Sort of.  This is mostly academic right now since you cannot host code-server at a base path without a trailing slash anyway because of the way browsers resolve relative paths (i.e. `/code` does not work because we would need the asset paths to be `./code/path/to/asset` which means we need to get `code` from somewhere which means we need to ask the user for the base path or we need to inject the asset paths on the client).  Current code-server does not work without a trailing slash (on the base path, it does work on our own routes like `/vscode` of course---I think I misspoke about this in Slack the other day so hopefully this clarifies).

There is one area we could potentially be breaking which is if existing users have a cookie at `/code` and we change the logic to use `/code/` then they can no longer log out.  But since we are also changing the cookie name that point is also moot.

I do still think we should avoid an arbitrary restriction though.  If one day we decide to support a slash-less base path it would be nice if the cookie logic already works.  So it is really just more about leaving support as broad as possible as long as there are no downsides in doing so.

Reply:
That is just this login portion though.  My comments about the logout stuff are still relevant though and I think need to be implemented.